### PR TITLE
sql: add metric for SQL connection latency

### LINKED ISSUE #66014: sql: add metric for SQL connection latency
We have metrics for number of SQL connections. It would be useful to also have observability on the latency of creating new SQL connections. Past incidents and customer debugging sessions have shown that this is an area worth measuring.

The metric should capture the duration from the time that CRDB receives the connection up until when it is ready to serve a SQL query. Specifically, it should include the time it takes to authenticate.

proposed name/metadata for this metric:

```
	MetaConnLatency = metric.Metadata{
		Name:        "sql.conn.latency",
		Help:        "Latency to establish a SQL connection",
		Measurement: "Nanoseconds",
		Unit:        metric.Unit_NANOSECONDS,
	}
```

### PR DESCRIPTION
**Pull request by mikecaat**
This adds a new metric "sql.conn.latency" to know the latency of
creating new SQL connections(#66014). It's useful for debugging sessions.

The metric specifies the duration between the time a new connection
is received and the time the connection is ready to serve a SQL query.
It also includes authentication time.

fixes #66014

Release note (ops change): Add a new metric to know the latency of
creating new SQL connection including authentication time.

Release note (ui change): Added a new chart showing the latency of
establishing a new SQL connection, including the time spent on
authentication.

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
thanks for this change! this is looking good, but I had a few comments

I'm not sure yet if we want to display it in the UI. let me find out.

if it is in the UI, we'd also need one more release note in the commit message:

```
Release note (ui): Added a new chart showing the latency of establishing a new
SQL connection, including the time spent on authentication.
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-:-McjgeHn3vkOY332ha0t:b-7mt408)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mikecaat)

___
*[pkg/sql/pgwire/conn.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McjcQpHAdqbdcqzCDs4:-McjcQpHAdqbdcqzCDs5:b8tyhrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f36785ec310ea12b57565efd68823bb15e979b/pkg/sql/pgwire/conn.go#L217)):*
> ```Go
> //
> // connStart is the time which the connection is activated. It's used for
> // calculating the establish connection latency until we can ready to
> ```

nit: this should say "It's used for calculating the establish connection latency until we *are* ready to serve a SQL query"
___
*[pkg/sql/pgwire/conn.go, line 403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McjceiY1nYnMQqGMyQY:-McjceiY1nYnMQqGMyQZ:b-ep2jul) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f36785ec310ea12b57565efd68823bb15e979b/pkg/sql/pgwire/conn.go#L403)):*
> ```Go
> 				authDone = true
> 
> 				// We count the connection establish latency until we can ready to
> ```

nit: should be "until we are ready"
___
*[pkg/sql/pgwire/server.go, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-Mcjg3OF7FrHtese4_WQ:-Mcjg3OF7FrHtese4_WR:b99ep09) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f36785ec310ea12b57565efd68823bb15e979b/pkg/sql/pgwire/server.go#L123)):*
> ```Go
> 	MetaConnLatency = metric.Metadata{
> 		Name:        "sql.conn.latency",
> 		Help:        "Latency to establish a SQL connection",
> ```

nit: make it say `"Latency to establish and authenticate a SQL connection"`
___
*[pkg/ts/catalog/chart_catalog.go, line 1991 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McjfEzN9JFFkMANtu9e:-McjfEzN9JFFkMANtu9f:b-5acszf) ([raw file](https://github.com/cockroachdb/cockroach/blob/20f36785ec310ea12b57565efd68823bb15e979b/pkg/ts/catalog/chart_catalog.go#L1991)):*
> ```Go
> 				Metrics: []string{
> 					"sql.conn.latency",
> 				},
> ```

Add another property: `AxisLabel: "Latency",`

**Review by knz — COMMENTED** [review]
Thanks for your change.

However, instead of a separate `connStart` variable passed as argument everywhere, please consider using a new field in the `conn` struct defined at the start of `conn.go`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-:-McjnToYDqZLtnEd073v:been3pf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mikecaat)

**Comment by knz** [comment]
cc @cockroachdb/server for education - here's an example of a good community PR.

**Comment by mikecaat** [comment]
Hi, @rafiss @knz 

Thanks for reviewing. I agreed with your comments. I fixed them.

> I'm not sure yet if we want to display it in the UI. let me find out.
> if it is in the UI, we'd also need one more release note in the commit message:

OK, I'll work if we need to add a new panel for it.

I thought it's enough that users can see the metrics in the "Custom Time Series Chart"
because there are not so many cases that this metric is useful for troubleshooting.
(Of course, there is a useful case, for example, fixing slow connectivity.)

But, the assumption may be wrong because I'm not familiar with CRDB...

Well, #66014 said the following. If you can, let me know the detail of the problems.
I think past experiences are the most important things to decide this.

>  Past incidents and customer debugging sessions have shown that this is an area worth measuring

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 5 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66625)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mikecaat)

___
*[pkg/sql/pgwire/conn.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McnRptBE4gY5MT1Lu8E:-McnRptBE4gY5MT1Lu8F:b-ngfm9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/conn.go#L175)):*
> ```Go
> 	c := &conn{
> 		conn:        netConn,
> 		sessionArgs: sArgs,
> ```

you can initialize `startTime` here
___
*[pkg/sql/pgwire/conn_test.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McnRUFi5xXz309vWQNU:-McnRUFi5xXz309vWQNV:bdltysr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/conn_test.go#L105)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	conn.startTime = timeutil.Now()
> ```

please remove this line
___
*[pkg/sql/pgwire/conn_test.go, line 1087 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McnRSX8Ey-nyUO0qITY:-McnRSX8Ey-nyUO0qITZ:bdltysr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/conn_test.go#L1087)):*
> ```Go
> 				nil,
> 			)
> 			conn.startTime = timeutil.Now()
> ```

please remove this line
___
*[pkg/sql/pgwire/server.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McnR_3N7krSrq9taX8j:-McnR_3O9lBop2DFb7Xo:bdltysr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/server.go#L687)):*
> ```Go
> 		ctx, conn, sArgs,
> 		reserved,
> 		connStart,
> ```

please remove this line

**Comment by mikecaat** [comment]
Hi, @knz 

Thanks for your comments. I have a question about the following comments.

> you can initialize startTime here

> pkg/sql/pgwire/server.go, line 687 at r2 (raw file):
> please remove this line

Is this means that startTime is the time when to call newConn()  like the following?

```
        c := &conn{
                conn:        netConn,
                sessionArgs: sArgs,
                metrics:     metrics,
+               startTime:   timeutil.Now(),
                rd:          *bufio.NewReader(netConn),
```

If so, "connStart" in ServeConn(), which is used for log event, and "startTime" 
in conn struct is different. Is it ok? I thought it's better to unify them.

If I'm misunderstanding something, please let me know.

**Review by rafiss — CHANGES_REQUESTED** [review]
I'll leave it up to knz to answer the question about `connStart` in ServeConn versus `startTime` in the conn struct.

About the UI: we discussed internally, and it would be desirable to have this chart be shown by default in the UI. I misunderstood your initial change; I thought your change added it as a default chart. Adding the chart is straightforward though: edit `pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/sql.tsx` and add the following. Add it above the `Service Latency: SQL Statements, 99th percentile` graph

```

    <LineGraph
      title="Connection Latency: 99th percentile"
      tooltip={`Over the last minute, this node established and authenticated 99% of connections within this time.`}
    >
      <Axis units={AxisUnits.Duration} label="latency">
        {_.map(nodeIDs, (node) => (
          <Metric
            key={node}
            name="cr.node.sql.conn.latency-p99"
            title={nodeDisplayName(nodesSummary, node)}
            sources={[node]}
            downsampleMax
          />
        ))}
      </Axis>
    </LineGraph>,

    <LineGraph
      title="Connection Latency: 90th percentile"
      tooltip={`Over the last minute, this node established and authenticated 90% of connections within this time.`}
    >
      <Axis units={AxisUnits.Duration} label="latency">
        {_.map(nodeIDs, (node) => (
          <Metric
            key={node}
            name="cr.node.sql.conn.latency-p90"
            title={nodeDisplayName(nodesSummary, node)}
            sources={[node]}
            downsampleMax
          />
        ))}
      </Axis>
    </LineGraph>,
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-:-McoVPbvEU7xYHXO_0UM:b-ts9umb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mikecaat)

___
*[pkg/sql/pgwire/conn.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McoOy_z63XE2GqtRCNl:-McoOya-At3C0efA606p:bhjuwz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/conn.go#L64)):*
> ```Go
> 	metrics     *ServerMetrics
> 
> 	// startTime is the time which the connection is activated.
> ```

I don't think "activated" is a clear term to use here. Can you make it say "startTime is the time when the connection attempt was first received by the server"

**Comment by mikecaat** [comment]
Hi, @rafiss

Thanks for your suggestions. I added the chart, fixed the comment, 
and updated the commit messages. Please review it.

**Review by knz — COMMENTED** [review]
> If so, "connStart" in ServeConn(), which is used for log event, and "startTime"
> in conn struct is different. Is it ok? I thought it's better to unify them.
> If I'm misunderstanding something, please let me know.

I have reviewed this into more details. In fact you are right, it is useful to reuse `connStart` properly.
However I do notice that the initialization of `connStart` is already "too late".
Can you please move it up, so that it becomes the first line at the start of `ServeConn()`?

Then I have updated my comments below.

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-:-Mctk5TJ0JRv0DrD8IcL:b-y0d79j)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mikecaat)

___
*[pkg/sql/pgwire/conn.go, line 171 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-MctkKy79NJmi-VHZc70:-MctkKy79NJmi-VHZc71:bo12z6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/conn.go#L171)):*
> ```Go
> 
> func newConn(
> 	netConn net.Conn, sArgs sql.SessionArgs, metrics *ServerMetrics, sv *settings.Values,
> ```

you can pass the `startTime` as argument to the conn constructor.
___
*[pkg/sql/pgwire/conn.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McnRptBE4gY5MT1Lu8E:-MctkN_YCyR3qUoezi3O:b75aix3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/conn.go#L175)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you can initialize `startTime` here
</blockquote></details>

you can initialize the field in the conn from the argument in the constructor.
___
*[pkg/sql/pgwire/conn.go, line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-MctkxmOBNMDCN8OCIDS:-MctkxmOBNMDCN8OCIDT:brv81ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/befa63454ba03f18c54546da1f1ae4059f546bcd/pkg/sql/pgwire/conn.go#L156)):*
> ```Go
> 	}
> 
> 	c := newConn(netConn, sArgs, &s.metrics, &s.execCfg.Settings.SV)
> ```

pass the connStart as argument to `newConn` here
___
*[pkg/sql/pgwire/conn_test.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McnRUFi5xXz309vWQNU:-MctkdoO22NxDaGzvZgC:b-bu7qc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/conn_test.go#L105)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

please remove this line
</blockquote></details>

you can still remove this line (initialization inside `waitForClientConn`)
___
*[pkg/sql/pgwire/conn_test.go, line 554 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-MctkZFoE-dsWg2GVnmO:-MctkZFoE-dsWg2GVnmP:bcvv1ld) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/conn_test.go#L554)):*
> ```Go
> 
> 	metrics := makeServerMetrics(sql.MemoryMetrics{} /* sqlMemMetrics */, metric.TestSampleInterval)
> 	pgwireConn := newConn(conn, sql.SessionArgs{ConnResultsBufferSize: 16 << 10}, &metrics, nil)
> ```

use `timeutil.Now` to initialize the argument to `newConn` here
___
*[pkg/sql/pgwire/server.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-McnR_3N7krSrq9taX8j:-MctkoL24obiMvFbH-X7:b5ow208) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2925332ab7902d4e651e8ec2c69dfb8f7b4755/pkg/sql/pgwire/server.go#L687)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

please remove this line
</blockquote></details>

here you can keep the argument (as discussed above)

**Comment by mikecaat** [comment]
Thanks for looking into the detail. 
I understood and updated the patch. Please check it.

Although it's not related to this issue directly,  I fixed a comment 
according to the actual code which  7db4e5cc forgot to change.

**Review by knz — APPROVED** [review]
:lgtm: I'm ok with this if Rafi is too!
thank you for your contribution

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-:-Mcy8tEeEgNLNcDAxaf5:b-vii6mn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mikecaat)

**Review by rafiss — APPROVED** [review]
looks great! thanks again

bors r=rafiss,knz

> Reviewed 1 of 5 files at r1, 2 of 4 files at r2, 1 of 2 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66625#-:-Mcyzj454fxbaaHMXxjJ:bsl1yc6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mikecaat)

**Comment by mikecaat** [comment]
Thank you for your kind help!

**Pull request merged**
